README.md
# vscode-translate

[![Current Version](http://vsmarketplacebadge.apphb.com/version/chun.vscode-translate.svg)](https://marketplace.visualstudio.com/items?itemName=chun.vscode-translate)
[![Install Count](http://vsmarketplacebadge.apphb.com/installs/chun.vscode-translate.svg)](https://marketplace.visualstudio.com/items?itemName=chun.vscode-translate)
[![Open Issues](http://vsmarketplacebadge.apphb.com/rating/chun.vscode-translate.svg)](https://marketplace.visualstudio.com/items?itemName=chun.vscode-translate)

## Translate languages.
<br />
press CTRL + t to turn on the translate then u can select the text and see the change of statusBar.
press CTRL + t to turn off the function. <br />

### It looks like:

![](https://raw.githubusercontent.com/jianzhichun/vscode-translate/master/img/translate_show.gif)

#### it will read http.proxy at user-settings to get data by proxy. 

#### and you can set the user-settings to change from language and target language. 

#### as for translation.api only google&baidu provided (default is baidu)

#### translation.languageDetection = true means do detection before language so that we could translate documents between from language and target language. (default is true) 

![](https://raw.githubusercontent.com/jianzhichun/vscode-translate/master/img/config.png)

### And you can press CTRL+SHIFT+R to replace selected-text to result after translation.

![](https://raw.githubusercontent.com/jianzhichun/vscode-translate/master/img/replace_show.gif)

## Language List (from google)

You can see [https://cloud.google.com/translate/docs/languages](https://cloud.google.com/translate/docs/languages)

| Language        | ISO-639-1 Code  
| ------------- |:-------------: |
| Afrikaans | af |
| Albanian | sq |
| Amharic | am |
| Arabic | ar |
| Armenian | hy |
| Azeerbaijani | az |
| Basque | eu |
| Belarusian | be |
| Bengali | bn |
| Bosnian | bs |
| Bulgarian | bg |
| Catalan | ca |
| Cebuano | ceb (ISO-639-2) |
| Chichewa | ny |
| Chinese (Simplified) | zh-CN (BCP-47) |
| Chinese (Traditional) | zh-TW (BCP-47) |
| Corsican | co |
| Croatian | hr |
| Czech | cs |
| Danish | da |
| Dutch | nl |
| English | en |
| Esperanto | eo |
| Estonian | et |
| Filipino | tl |
| Finnish | fi |
| French | fr |
| Frisian | fy |
| Galician | gl |
| Georgian | ka |
| German | de |
| Greek | el |
| Gujarati | gu |
| Haitian Creole | ht |
| Hausa | ha |
| Hawaiian | haw (ISO-639-2) |
| Hebrew | iw |
| Hindi | hi |
| Hmong | hmn (ISO-639-2) |
| Hungarian | hu |
| Icelandic | is |
| Igbo | ig |
| Indonesian | id |
| Irish | ga |
| Italian | it |
| Japanese | ja |
| Javanese | jw |
| Kannada | kn |
| Kazakh | kk |
| Khmer | km |
| Korean | ko |
| Kurdish | ku |
| Kyrgyz | ky |
| Lao | lo |
| Latin | la |
| Latvian | lv |
| Lithuanian | lt |
| Luxembourgish | lb |
| Macedonian | mk |
| Malagasy | mg |
| Malay | ms |
| Malayalam | ml |
| Maltese | mt |
| Maori | mi |
| Marathi | mr |
| Mongolian | mn |
| Burmese | my |
| Nepali | ne |
| Norwegian | no |
| Pashto | ps |
| Persian | fa |
| Polish | pl |
| Portuguese | pt |
| Punjabi | ma |
| Romanian | ro |
| Russian | ru |
| Samoan | sm |
| Scots Gaelic | gd |
| Serbian | sr |
| Sesotho | st |
| Shona | sn |
| Sindhi | sd |
| Sinhala | si |
| Slovak | sk |
| Slovenian | sl |
| Somali | so |
| Spanish | es |
| Sundanese | su |
| Swahili | sw |
| Swedish | sv |
| Tajik | tg |
| Tamil | ta |
| Telugu | te |
| Thai | th |
| Turkish | tr |
| Ukrainian | uk |
| Urdu | ur |
| Uzbek | uz |
| Vietnamese | vi |
| Welsh | cy |
| Xhosa | xh |
| Yiddish | yi |
| Yoruba | yo |
| Zulu | zu |


## Contribution.

[https://github.com/jianzhichun/vscode-translate](https://github.com/jianzhichun/vscode-translate)

## Thanks

[Jesse-Deng](https://github.com/Jesse-Deng)

**Enjoy!**

### 2016/8/28
    init 0.0.1
### 2016/8/29
    init 0.0.2 add proxy support 
### 2016/8/29
    update README.md
### 2017/3/4 
    update to 0.0.4
    add others translation support
    add replace support
### 2017/3/4
    update to 0.0.5
    add google translation optional
### 2017/3/9
    update to 0.1.0
    add language detection support

CHANGELOG.md
# Change Log
All notable changes to the "vscode-translate" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

out/src/extension.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode_1 = require("vscode");
const WebRequest = require("web-request");
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
let isActive = false, translateResult = '', google2baidu = {
    en: 'en',
    th: 'th',
    ru: 'ru',
    pt: 'pt',
    el: 'el',
    nl: 'nl',
    pl: 'pl',
    bg: 'bul',
    et: 'est',
    da: 'dan',
    fi: 'fin',
    cs: 'cs',
    ro: 'rom',
    sl: 'slo',
    sv: 'swe',
    hu: 'hu',
    de: 'de',
    it: 'it',
    zh: 'zh',
    'zh-CN': 'zh',
    'zh-TW': 'cht',
    'zh-HK': 'yue',
    ja: 'jp',
    ko: 'kor',
    es: 'spa',
    fr: 'fra',
    ar: 'ara'
};
function activate(context) {
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    context.subscriptions.push(new Translate());
    context.subscriptions.push(vscode_1.commands.registerCommand('extension.translateon', () => {
        if (!isActive) {
            vscode_1.window.showInformationMessage('Translate on!');
            isActive = true;
        }
        else {
            vscode_1.window.showInformationMessage('Translate off!');
            isActive = false;
        }
    }));
    context.subscriptions.push(vscode_1.commands.registerCommand('extension.translateReplace', () => {
        if (!isActive) {
            return;
        }
        var editor = vscode_1.window.activeTextEditor, selection = editor.selection;
        editor.edit(edit => edit.replace(selection, translateResult));
    }));
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() {
}
exports.deactivate = deactivate;
class Translate {
    constructor() {
        if (!this.statusBarItem) {
            this.statusBarItem = vscode_1.window.createStatusBarItem(vscode_1.StatusBarAlignment.Left);
        }
        let subscriptions = [];
        vscode_1.window.onDidChangeTextEditorSelection(this.updateTranslate, this, subscriptions);
        this.disposable = vscode_1.Disposable.from(...subscriptions);
        this.updateTranslate();
        this.statusBarItem.show();
    }
    updateTranslate() {
        var cfg = vscode_1.workspace.getConfiguration(), proxy = String(cfg.get("http.proxy")), api = String(cfg.get("translation.api")), targetLanguage = String(cfg.get("translation.targetLanguage")), fromLanguage = String(cfg.get("translation.fromLanguage")), languageDetection = Boolean(cfg.get("translation.languageDetection"));
        let editor = vscode_1.window.activeTextEditor;
        if (!isActive || !editor) {
            this.statusBarItem.hide();
            return;
        }
        let doc = editor.document, str = doc.getText(editor.selection).trim();
        if (str.trim() == '')
            return;
        setTimeout(() => {
            languageDetection
                ? this.languageDetection(str, fromLanguage).then((isReverse) => {
                    if (isReverse) {
                        [fromLanguage, targetLanguage] = [targetLanguage, fromLanguage];
                    }
                    this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
                })
                : this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
        }, 1000);
    }
    dotranslate(str, proxy, api, targetLanguage, fromLanguage) {
        var statusBarItem = this.statusBarItem;
        var translateStr = api == 'baidu' ? this.baiduTranslate(str, google2baidu[targetLanguage], google2baidu[fromLanguage]) : this.googleTranslate(str, targetLanguage, fromLanguage);
        WebRequest.get(translateStr, { "proxy": proxy }).then(function (TResult) {
            var rs = '', res = JSON.parse(TResult.content.toString());
            if (api == 'baidu') {
                if (res.error)
                    return;
                rs = res.trans_result.data[0].dst;
            }
            else {
                var result = [];
                res.sentences.forEach(function (v) {
                    result.push(v.trans);
                });
                rs = result.join(',');
            }
            statusBarItem.text = translateResult = rs;
            statusBarItem.show();
        });
    }
    languageDetection(str, fromLanguage) {
        return WebRequest.post('http://fanyi.baidu.com/langdetect', { "formData": { "query": str }, "timeout": 500 }).then(function (result) {
            var res = JSON.parse(result.content);
            if (res.error || res.lan == fromLanguage || fromLanguage == '') {
                return false;
            }
            else {
                return true;
            }
        }, function () {
            return false;
        });
    }
    baiduTranslate(str, targetLanguage, fromLanguage) {
        return 'http://fanyi.baidu.com/v2transapi?query=' + str + (fromLanguage ? '&from=' + fromLanguage : '') + (targetLanguage ? '&to=' + targetLanguage : '');
    }
    googleTranslate(str, targetLanguage, fromLanguage) {
        return 'https://translate.google.cn/translate_a/single?client=gtx&sl=' + (fromLanguage || 'auto') + '&tl=' + (targetLanguage || 'auto') + '&hl=zh-CN&dt=t&dt=bd&ie=UTF-8&oe=UTF-8&dj=1&source=icon&q=' + str;
    }
    dispose() {
        this.statusBarItem.dispose();
        this.disposable.dispose();
    }
}
//# sourceMappingURL=extension.js.map

src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import { window, workspace, commands, Disposable, ExtensionContext, StatusBarAlignment, StatusBarItem, TextDocument } from 'vscode';
import * as WebRequest from 'web-request';
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
let isActive = false
    , translateResult = ''
    , google2baidu = {
        en: 'en',
        th: 'th',
        ru: 'ru',
        pt: 'pt',
        el: 'el',
        nl: 'nl',
        pl: 'pl',
        bg: 'bul',
        et: 'est',
        da: 'dan',
        fi: 'fin',
        cs: 'cs',
        ro: 'rom',
        sl: 'slo',
        sv: 'swe',
        hu: 'hu',
        de: 'de',
        it: 'it',
        zh: 'zh',
        'zh-CN': 'zh',
        'zh-TW': 'cht',
        'zh-HK': 'yue',
        ja: 'jp',
        ko: 'kor',
        es: 'spa',
        fr: 'fra',
        ar: 'ara'
    };


export function activate(context: ExtensionContext) {

    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    context.subscriptions.push(new Translate());
    context.subscriptions.push(commands.registerCommand('extension.translateon', () => {
        if (!isActive) {
            window.showInformationMessage('Translate on!');
            isActive = true;
        } else {
            window.showInformationMessage('Translate off!');
            isActive = false;
        }
    }));
    context.subscriptions.push(commands.registerCommand('extension.translateReplace', () => {
        if (!isActive) {
            return;
        }
        var editor = window.activeTextEditor
            , selection = editor.selection;
        editor.edit(edit =>
            edit.replace(selection, translateResult)
        );
    }));

}

// this method is called when your extension is deactivated
export function deactivate() {
}
class Translate {

    private statusBarItem: StatusBarItem;
    private disposable: Disposable;
    constructor() {
        if (!this.statusBarItem) {
            this.statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);
        }
        let subscriptions: Disposable[] = [];
        window.onDidChangeTextEditorSelection(this.updateTranslate, this, subscriptions);
        this.disposable = Disposable.from(...subscriptions);
        this.updateTranslate();
        this.statusBarItem.show();
    }
    public updateTranslate() {
        var cfg = workspace.getConfiguration()
            , proxy = String(cfg.get("http.proxy"))
            , api = String(cfg.get("translation.api"))
            , targetLanguage = String(cfg.get("translation.targetLanguage"))
            , fromLanguage = String(cfg.get("translation.fromLanguage"))
            , languageDetection = Boolean(cfg.get("translation.languageDetection"));
        let editor = window.activeTextEditor;
        if (!isActive || !editor) {
            this.statusBarItem.hide();
            return;
        }

        let doc = editor.document
            , str = doc.getText(editor.selection).trim();
        if (str.trim() == '')
            return
        setTimeout(() => {
            languageDetection
                ? this.languageDetection(str, fromLanguage).then((isReverse) => {
                    if (isReverse) {
                        [fromLanguage, targetLanguage] = [targetLanguage, fromLanguage];
                    }
                    this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
                })
                : this.dotranslate(encodeURIComponent(str), proxy, api, targetLanguage, fromLanguage);
        }, 1000);


    }
    private dotranslate(str, proxy, api, targetLanguage, fromLanguage) {
        var statusBarItem = this.statusBarItem;
        var translateStr = api == 'baidu' ? this.baiduTranslate(str, google2baidu[targetLanguage], google2baidu[fromLanguage]) : this.googleTranslate(str, targetLanguage, fromLanguage);
        WebRequest.get(translateStr, { "proxy": proxy }).then(function (TResult) {
            var rs = ''
                , res = JSON.parse(TResult.content.toString());
            if (api == 'baidu') {
                if (res.error) return;
                rs = res.trans_result.data[0].dst;
            } else {
                var result = []
                res.sentences.forEach(function (v) {
                    result.push(v.trans)
                })
                rs = result.join(',');
            }
            statusBarItem.text = translateResult = rs;
            statusBarItem.show();
        });
    }
    private languageDetection(str, fromLanguage) {
        return WebRequest.post('http://fanyi.baidu.com/langdetect', { "formData": { "query": str }, "timeout": 500 }).then(function (result) {
            var res = JSON.parse(result.content);
            if (res.error || res.lan == fromLanguage || fromLanguage == '') {
                return false;
            } else {
                return true;
            }
        }, function () {
            return false;
        });
    }
    private baiduTranslate(str, targetLanguage, fromLanguage) {
        return 'http://fanyi.baidu.com/v2transapi?query=' + str + (fromLanguage ? '&from=' + fromLanguage : '') + (targetLanguage ? '&to=' + targetLanguage : '');

    }
    private googleTranslate(str, targetLanguage, fromLanguage) {
        return 'https://translate.google.cn/translate_a/single?client=gtx&sl=' + (fromLanguage || 'auto') + '&tl=' + (targetLanguage || 'auto') + '&hl=zh-CN&dt=t&dt=bd&ie=UTF-8&oe=UTF-8&dj=1&source=icon&q=' + str;
    }

    dispose() {
        this.statusBarItem.dispose();
        this.disposable.dispose();
    }
}
